scripts/install/install-helper-functions.sh

```diff
@@ -0,0 +1,148 @@
+# This file is to be sourced into another to add helper functions, rather than
+# executed directly.
+
+#-------------------------------------------------------------------------------
+# Logging support (nascent; I want to add more control and a log file).
+
+# Print a separator bar of # characters.
+function echo_bar() {
+  local terminal_width
+  if [ -n "${TERM}" ] && [ -t 0 ] ; then
+    if [[ -n "$(which resize)" ]] ; then
+      terminal_width="$(resize 2>/dev/null | grep COLUMNS= | cut -d= -f2)"
+    elif [[ -n "$(which stty)" ]] ; then
+      terminal_width="$(stty size 2>/dev/null | cut '-d ' -f2)"
+    fi
+  fi
+  printf "%${terminal_width:-80}s\n" | tr ' ' '#'
+}
+
+function echo_running_sudo() {
+  if [ "$(id -u -n)" == "root" ] ; then
+    echo "Running $1"
+  else
+    echo "Running sudo $1; you may be prompted for your password."
+  fi
+}
+
+#-------------------------------------------------------------------------------
+# Misc helper functions.
+
+# Execute the args (a command) as root, either using sudo or directly if already
+# running as root.
+function my_sudo() {
+  if [ "$(id -u -n)" == "root" ] ; then
+    "$@"
+  else
+    (set -x ; sudo "$@")
+  fi
+}
+
+# Get the type of the first arg, i.e. shell function, executable, etc.
+# For more info: https://ss64.com/bash/type.html
+#            or: https://bash.cyberciti.biz/guide/Type_command
+function safe_type() {
+  type -t "${1}" || /bin/true
+}
+
+# Print the disk location of the first arg.
+function safe_which() {
+  type -p "${1}" || /bin/true
+}
+
+# Does the first arg start with the second arg?
+function beginswith() { case "${1}" in "${2}"*) true;; *) false;; esac; }
+
+#-------------------------------------------------------------------------------
+# Helpers for finding the APT caching proxy.
+
+# 
+function proc_net_route_addr() {
+  if [ ! -e /proc/net/route ] ; then
+    return
+  fi
+  # This can be done with awk, but that is then much harder to debug.
+  # /proc/net/route has contents like this:
+  #    Iface Destination Gateway  Flags RefCnt Use Metric Mask     MTU Window IRTT                                                       
+  #    eth0  00000000    010011AC 0003  0      0   0      00000000 0   0      0                                                                               
+  # We're looking to find a line like the second one, with zero for the destination
+  # and the Gateway is the 32-bit hex value for the IPv4 address of the docker host,
+  # where the last two characters represent the leading (high) bits of the address.
+  local -r -a words=( $(grep -E -i '^[a-z][a-z0-9]+\s00000000\s[0-9a-f][0-9a-f][0-9a-f][0-9a-f]' /proc/net/route ) )
+  local -r gateway="${words[2]}"
+  local -r a="0x${gateway:6:2}"
+  local -r b="0x${gateway:4:2}"
+  local -r c="0x${gateway:2:2}"
+  local -r d="0x${gateway:0:2}"
+  printf "%d.%d.%d.%d\n" "${a}" "${b}" "${c}" "${d}"
+}
+
+# Print the first IP on the route to 8.8.8.8, i.e. to Google's public DNS server?
```

Comment:
Don't think this is supposed to be a question.

Reply:
Right!